### PR TITLE
MSR: Remove use of reflection in UCOs of generic types

### LINKED ISSUE #18356: Research injecting an interface to resolve open generic members and types to closed ones in the managed static registrar
Ref: https://dotnetfiddle.net/orpEt8
Ref: https://github.com/dotnet/runtime/issues/86649#issuecomment-1564461888

This would certainly speed up calling generic types/members from Objective-C, and might also have a size benefit.

```cs
using System;
using System.Runtime.InteropServices;

var inst = new GenericNSObject<Dummy>();
var handle = GCHandle.Alloc(inst);

var res = GenericNSObject_1__Registrar_Callbacks__.DoSomething((IntPtr)handle, 12, true);
Console.WriteLine($"res: {res}");

var res2 = GenericNSObject_1__Registrar_Callbacks__.DoSomething((IntPtr)handle, 24, false);
Console.WriteLine($"res2: {res2}");

class Dummy {}

// app code
public partial class GenericNSObject<T> : NSObject {
    // [Export("blah")]
    public int DoSomething(int p0, bool p1) {
        Console.WriteLine($"{typeof(T)} - {p0}, {p1}");
        return p1 ? p0 : -p0;
    }
}

// generated code
internal interface GenericNSObject_1__GenericsProxy {
    int DoSomething_Proxy(int p0, bool p1);
}

partial class GenericNSObject<T> : GenericNSObject_1__GenericsProxy {
    public int DoSomething_Proxy(int p0, bool p1) => DoSomething(p0, p1);
}

internal static class GenericNSObject_1__Registrar_Callbacks__ {
    // [UnmanagedCallersOnly(EntryPoint = "__calback_123_GenericNSObject_1__DoSomething")]
    public static int DoSomething(IntPtr handle, int p0, bool p1) {
        var obj = Runtime.GetNSObject<GenericNSObject_1__GenericsProxy>(handle);
        return obj.DoSomething_Proxy(p0, p1);
    }
}

public class NSObject
{
}

public class Runtime
{
    public static T GetNSObject<T>(IntPtr handle)
        => (T)GCHandle.FromIntPtr(handle).Target!;
}
```

Credit to @simonrozsival for this idea.

### PR DESCRIPTION
**Pull request by simonrozsival**
Closes #18356

In the static UnmanagedCallersOnly methods we don't know the generic parameters of the type we're working with and we need to use this trick to be able to call methods on the generic type without using reflection. When we call a non-generic interface method implemented on a generic class, the .NET runtime will resolve the generic parameters for us. In the implementation of the interface method, we can simply use the generic parameters and generate the same code we usually generate in the UnmanagedCallersOnly callback method.

This is an example of the code we generate in addition to user code:

```csharp
internal interface __IRegistrarGenericTypeProxy__CustomNSObject_1__
{
	void __IRegistrarGenericTypeProxy__CustomNSObject_1____SomeMethod (IntPtr p0);
}

public class CustomNSObject<T> : NSObject, __IRegistrarGenericTypeProxy__CustomNSObject_1__
	where T : NSObject
{
	[Export ("someMethod:")]
	public void SomeMethod (T someInput)
	{
		// ...
	}

	// generated implementation of the proxy interface:
	public void __IRegistrarGenericTypeProxy__CustomNSObject_1____SomeMethod (IntPtr sel, IntPtr p0, IntPtr* exception_gchandle)
	{
		try {
			var obj0 = (T) Runtime.GetNSObject<T> (p0);
			SomeMethod (obj0);
		} catch (Exception ex) {
			*exception_gchandle = Runtime.AllocGCHandle (ex);
		}
	}

	// generated registrar callbacks:
	private static class __Registrar_Callbacks__
	{
		[UnmanagedCallersOnly (EntryPoint = "_callback_1_CustomNSObject_1_SomeMethod")]
		public unsafe static void callback_1_CustomNSObject_1_SomeMethod (IntPtr pobj, IntPtr sel, IntPtr p0, IntPtr* exception_gchandle)
		{
			var proxy = (__IRegistrarGenericTypeProxy__CustomNSObject_1__)Runtime.GetNSObject (pobj);
			proxy.__IRegistrarGenericTypeProxy__CustomNSObject_1____SomeMethod (sel, p0, exception_gchandle);
		}
	}
}
```
```csharp
// regular non-generic class for comparison:
public class CustomNSObject : NSObject
	where T : NSObject
{
	[Export ("someMethod:")]
	public void SomeMethod (NSSet someInput)
	{
		// ...
	}
	
	private static class __Registrar_Callbacks__
	{
		[UnmanagedCallersOnly (EntryPoint = "_callback_1_CustomNSObject_1_SomeMethod")]
		public unsafe static void callback_1_CustomNSObject_1_SomeMethod (IntPtr pobj, IntPtr sel, IntPtr p0, IntPtr* exception_gchandle)
		{
			try {
				NSSet obj0 = Runtime.GetNSObject<NSSet> (p0);
				SomeMethod (obj0);
			} catch (Exception ex) {
				*exception_gchandle = Runtime.AllocGCHandle (ex);
			}
		}
	}
}
```

### CODE REVIEW

File: tools/dotnet-linker/Steps/ManagedRegistrarStep.cs [hunk 1]
```diff
@@ -990,6 +948,20 @@ bool IsOpenType (TypeReference tr)
...
 
+		TypeReference ReplaceOpenTypeIfNeeded (TypeReference type, TypeReference replacementType)
+		{
+			if (IsOpenType (type)) {
+				if (type is ArrayType arrayType) {
+					var elementType = ReplaceOpenTypeIfNeeded (arrayType.ElementType, replacementType);
+					return new ArrayType (elementType, arrayType.Rank);
+				}
+
+				return replacementType;
```
Comment (rolfbjarne) [inline tools/dotnet-linker/Steps/ManagedRegistrarStep.cs:959]: What about out/ref parameters?

File: tools/dotnet-linker/Steps/ManagedRegistrarStep.cs [hunk 2]
```diff
@@ -516,6 +490,40 @@ void CreateUnmanagedCallersMethod (MethodDefinition method, AssemblyTrampolineIn
 			eh.HandlerEnd = (Instruction) leaveEHInstruction.Operand;
 		}
 
+		MethodDefinition CreateGenericsProxyMethod (MethodDefinition method, MethodDefinition callback, List<TypeDefinition> proxyInterfaces)
+		{
+			var proxyInterfaceName = $"__IRegistrarGenericTypeProxy__{Sanitize (method.DeclaringType.FullName)}__";
```
Comment (rolfbjarne) [inline tools/dotnet-linker/Steps/ManagedRegistrarStep.cs:495]: Can you add a comment here with an example in C# of the code you're generating?

File: tools/dotnet-linker/Steps/ManagedRegistrarStep.cs [hunk 3]
```diff
@@ -516,6 +490,40 @@ void CreateUnmanagedCallersMethod (MethodDefinition method, AssemblyTrampolineIn
...
+			TypeDefinition? proxyInterface = proxyInterfaces.SingleOrDefault (v => v.Name == proxyInterfaceName && v.Namespace == "ObjCRuntime");
+			if (proxyInterface is null) {
+				proxyInterface = new TypeDefinition ("ObjCRuntime", proxyInterfaceName, TypeAttributes.Public | TypeAttributes.Interface | TypeAttributes.Abstract);
```
Comment (rolfbjarne) [inline tools/dotnet-linker/Steps/ManagedRegistrarStep.cs:498]: I don't think the interface needs to be public, it can be internal.

**Review by rolfbjarne — COMMENTED** [review]
There's one failing test: it's failing because an exception isn't thrown when calling a method taking a generic parameter T instantiated with NSSet is passed NSObject: https://github.com/xamarin/xamarin-macios/blob/e3088879a619ba2af5d5d739d1d212243e810f24/tests/monotouch-test/ObjCRuntime/RegistrarTest.cs#L1351

File: tools/dotnet-linker/Steps/ManagedRegistrarStep.cs (hunk 1 above)
Comment (simonrozsival) [reply tools/dotnet-linker/Steps/ManagedRegistrarStep.cs]: I added the implementation in the last commit a moment ago.

**Comment by marek-safar** [comment]
> 	// generated registrar callbacks:
	private static class __Registrar_Callbacks__

Just wondering why is it nested class and not high-level type with X generated methods where we reuse the ones with the same signature?

**Comment by filipnavara** [comment]
> Just wondering why is it nested class and not high-level type with X generated methods where we reuse the ones with the same signature?

[Cannot speak for the authors but first time I read the MSR output I assumed] it's because the exported methods can be private/internal and nested classes naturally have access to those methods.

**Comment by simonrozsival** [comment]
> Just wondering why is it nested class and not high-level type with X generated methods where we reuse the ones with the same signature?

I'm not sure how that would work for `__Registrar_Callbacks__` 🤔 Maybe Rolf has an idea how we could reuse UCOs with the same signature. My understanding is the same as Filip's in that the nesting helps getting around visibility issues.

I think it would be possible to share the proxy interfaces that are generated in this PR although it might be tricky to get right and I'm not sure how much we could bring the app size down if we did that.

**Comment by simonrozsival** [comment]
@rolfbjarne I fixed the bug that caused the `GenericClassWithUnrelatedGenericDelegate`. Please re-run the tests one more time.

**Review by rolfbjarne — APPROVED** [review]

**Comment by rolfbjarne** [comment]
Great work!

**Pull request merged**